Experimental/src/Ranges/Range.cs

```diff
@@ -0,0 +1,713 @@
+        }
+
+        public bool Equals(Range<TValue> other)
+        {
+            return _options == other._options
+                   && EqualityComparer<TValue>.Default.Equals(End, other.End) &&
+                   EqualityComparer<TValue>.Default.Equals(Start, other.Start);
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (ReferenceEquals(null, obj)) return false;
+            return obj is Range<TValue> && Equals((Range<TValue>) obj);
+        }
+
+        public override int GetHashCode()
+        {
+            unchecked
+            {
+                var hashCode = (int) _options;
+                hashCode = (hashCode * 397) ^ EqualityComparer<TValue>.Default.GetHashCode(End);
+                hashCode = (hashCode * 397) ^ EqualityComparer<TValue>.Default.GetHashCode(Start);
+                return hashCode;
+            }
+        }
+
+        public Range<TValue> SetStart(TValue newStart, bool? include = null)
+        {
+            return new Range<TValue>(newStart, End,
+                ((include ?? IncludeStart) ? RangeOptions.IncludingStart : RangeOptions.None) |
+                (IncludeEnd ? RangeOptions.IncludingEnd : RangeOptions.None));
+        }
+
+        public Range<TValue> SetEnd(TValue newEnd, bool? include = null)
+        {
+            return new Range<TValue>(Start, newEnd,
+                (IncludeStart ? RangeOptions.IncludingStart : RangeOptions.None) |
+                ((include ?? IncludeEnd) ? RangeOptions.IncludingEnd : RangeOptions.None));
+        }
+
+        public Range<TValue> Union(Range<TValue> other)
+        {
+            var current = this;
+
+            if (current.IsEmpty)
+                return other;
+
+            if (other.IsEmpty)
+                return this;
+
+            RangeValue newStart;
+            RangeValue newEnd;
+
+            newStart = RangeValue.MinValue(other.StartValue, current.StartValue, false);
+            newEnd = RangeValue.MaxValue(other.EndValue, current.EndValue, false);
+
+            return new Range<TValue>(newStart, newEnd);
+        }
+
+        public bool Contains(TValue value)
+        {
+            return Contains(value, true);
+        }
+
+        private bool Contains(TValue value, bool included)
+        {
+            if (IsEmpty)
+                return false;
+
+            bool result = true;
+
+            if (HasStart)
+            {
+                var compare = Start.CompareTo(value);
+                result = compare < 0 || compare == 0 && IncludeStart && included;
+            }
+
+            if (result && HasEnd)
+            {
+                var compare = End.CompareTo(value);
+                result = compare > 0 || compare == 0 && IncludeEnd && included;
+            }
+
+            return result;
+        }
+
+        public bool IsAdjastent(Range<TValue> other)
+        {
+            var current = this;
+            if (current.HasEnd && other.HasStart)
```

Comment:
Please, move this check to a separate private method and just call it twice for (this, other) and (other, this)